### ISSUE TITLE
Doesn't work with Docker Hub access tokens

### ISSUE THREAD

**Issue by gotson** [peter-evans/dockerhub-description#10]
Hello,

i tried using this with the recently introduced DockerHub access tokens, but it doesn't work. I swapped my credentials for my own Docker Hub and it works again.

If i check the Docker Hub access tokens page, i see that the access token was never used.

When using the access token i get an error 403 on the PATCH request.

**Comment by peter-evans** [comment]
Yes, unfortunately, DockerHub don't support access tokens for this API at the moment. There is a note about this in the readme.

> **Note**: Docker Hub [Personal Access Tokens](https://docs.docker.com/docker-hub/access-tokens/) cannot be used as they are not supported by the API. See [here](https://github.com/docker/hub-feedback/issues/1927) and [here](https://github.com/docker/hub-feedback/issues/1914) for further details. Unfortunately, this means that enabling the new 2FA feature on Docker Hub will prevent the action from working.

https://github.com/peter-evans/dockerhub-description/blob/master/README.md#required-environment-variables

**Comment by gotson** [comment]
My bad, i didn't read the documentation since they introduced the access tokens. Thanks!

**Comment by peter-evans** [comment]
No worries. I hope the open up the API for use with tokens at some point!

**Comment by cooperlees** [comment]
O damn, I've hit this. Has anyone found somewhere we can request support for this?

**Comment by peter-evans** [comment]
@cooperlees This is the issue tracking the roadmap feature to allow tokens to be used. https://github.com/docker/roadmap/issues/115

**Comment by cooperlees** [comment]
Could we possibly leave this open so people see it's a known issue? It will possibly help stop duplicate issues being logged.

**Comment by LongLiveCHIEF** [comment]
It's not like the tokens are any different than passwords anyways. Tokens have absolutely no scope attached to them.  Docker has needed to step up their game with this stuff for _years_.

**Comment by craighurley** [comment]
@peter-evans , docker hub started supporting scoped tokens; I've looked at the docs you original linked to and it no longer states that "they are not supported by the API"... does that unblock this issue?

https://www.docker.com/blog/level-up-security-with-scoped-access-tokens/
https://github.com/docker/roadmap/issues/115

**Comment by peter-evans** [comment]
@craighurley The scoped tokens feature they have just announced is, unfortunately, only provided for Pro and Team accounts. I have a free account so I tested the action again with a `Read, Write, Delete` scoped token (the only option available to free accounts) and it still doesn't work. I don't think this update fixes the issue because the API in use by this action still doesn't seem to support access tokens altogether.

There is a possibility that they have only enabled access tokens for this API if you are on a Pro or Team plan. If anyone is on a Pro or Team plan and can test it to confirm, that would be helpful.

**Comment by modem7** [comment]
> @craighurley The scoped tokens feature they have just announced is, unfortunately, only provided for Pro and Team accounts. I have a free account so I tested the action again with a `Read, Write, Delete` scoped token (the only option available to free accounts) and it still doesn't work. I don't think this update fixes the issue because the API in use by this action still doesn't seem to support access tokens altogether.
> 
> There is a possibility that they have only enabled access tokens for this API if you are on a Pro or Team plan. If anyone is on a Pro or Team plan and can test it to confirm, that would be helpful.

https://github.com/docker/roadmap/issues/115#issuecomment-891694974

**Comment by spaquet** [comment]
Also, when enabling 2FA there is no way to use username/password to login from the CLI.

**Comment by kerphi** [comment]
I did not catch if the DOCKERHUB_TOKEN can now be used by peter-evans/dockerhub-description ?

I just see this but it was published on fev 2020 so maybe it's obsolete?
> Docker Hub [Personal Access Tokens](https://docs.docker.com/docker-hub/access-tokens/) cannot be used as they are not supported by the API.

**Comment by peter-evans** [comment]
@kerphi Docker Hub Personal Access Tokens still do not work. The request for this feature seems to have moved to a new issue here: https://github.com/docker/hub-tool/issues/172

**Comment by wolfgangwalther** [comment]
It seems like the Dockerhub API should support Personal Access Tokens now: https://github.com/docker/hub-feedback/issues/2127

My jobs are still failing, though - so I guess some changes need to be made to this action, too?

**Comment by christian-korneck** [comment]
is this API call working with a PAT for anyone here? I tried but [it didn't seem to work](https://github.com/docker/hub-feedback/issues/2127#issuecomment-1152831106). If so, I'd appreciate a pointer into the right direction.

**Comment by peter-evans** [comment]
Docker Hub have [announced access token support](https://github.com/docker/hub-feedback/issues/2127#issuecomment-1183192693) has been rolled out. I've not had chance to test it yet, but it looks like this update could be the one to fix this issue.

**Comment by simmel** [comment]
> Docker Hub have [announced access token support](https://github.com/docker/hub-feedback/issues/2127#issuecomment-1183192693) has been rolled out. I've not had chance to test it yet, but it looks like this update could be the one to fix this issue.

Can confirm that using PAT now works.

**Comment by peter-evans** [comment]
Thanks for confirming! I think we can close this now.

**Comment by jaydrogers** [comment]
I hate to wake this old thread up, but I figured it was the best place to keep this discussion organized. Today I am running into this error. No changes were made on my end and it just started to appear this morning (I have a job run every Tuesday to automatically update).

## Error Message

```
Error: Bad Request
```

#### Full screenshot
<img width="1264" alt="image" src="https://user-images.githubusercontent.com/3174134/218785772-3b9ba834-c40f-4807-9840-3a0f177eca3e.png">

## Attempted fix (same failing result)
- Delete my PAT
- Re-create it with read, write, delete
- Re-run the job

## Possible error to help debug?
```
Run peter-evans/dockerhub-description@v3
  with:
    username: ***
    password: ***
    repository: serversideup/php
    short-description: Production-ready Docker images for PHP. Optimized for Laravel, WordPress, and more!
##[debug]Inputs: {
##[debug]  username: '***',
##[debug]  password: '***',
##[debug]  repository: 'serversideup/php',
##[debug]  shortDescription: 'Production-ready Docker images for PHP. Optimized for Laravel, WordPress, and more!',
##[debug]  readmeFilepath: './README.md'
##[debug]}
Acquiring token
::add-mask::***
Sending PATCH request
##[debug]Error: Bad Request
##[debug]    at /home/runner/work/_actions/peter-evans/dockerhub-description/v3/dist/index.js:79:23
##[debug]    at processTicksAndRejections (node:internal/process/task_queues:96:5)
Error: Bad Request
##[debug]Node Action run completed with exit code 1
##[debug]Finishing: push README to Dockerhub
```

## My source code
https://github.com/serversideup/docker-php/blob/649c0e215a82512554ce7e4f616f099dd2af8b3c/.github/workflows/publish_docker-images-production.yml#L59-L65

## Other notes & thanks
- I have another repository running the same syntax and it seems to be fine 🤷‍♂️

I'm very grateful for this project! Thank you for putting this together. If anyone has any ideas, it would be very appreciated 🙌

**Comment by simmel** [comment]
@jaydrogers I just tried to update and it worked see https://github.com/simmel/vanilla-imap-baby/actions/runs/4177535798/jobs/7235219564#step:8:1

Try again? Might have just been Docker Hubs API that had an issue?

**Comment by jaydrogers** [comment]
Thanks @simmel!

I just tried it again, and I am getting the same results. I can move this into a new issue if it's more appropriate 😃

<img width="1330" alt="image" src="https://user-images.githubusercontent.com/3174134/218851027-4c90a05e-55f7-4a45-8c30-e9a6e0a0f535.png">

**Comment by jonasbn** [comment]
Hi @jaydrogers 

Here are my 2 cents.

The error hints at a bad HTTP request (`400`), ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

- Is there anyway to get to see the original request and response?
- Evaluating what the variable parts are - does the username or password contain anything that could render the request invalid?
- Are there other variable parameters
- Are the credentials the same as for the other repository where it works?

**Comment by jonasbn** [comment]
Hi @jaydrogers 

And 2 cents more

Just a last thing, have you tried the running the Docker image outside of the context of GHA, as outlined in the README?

**Comment by jaydrogers** [comment]
Thanks @jonasbn! Brilliant method with the Docker testing. 👍

I attempted to run what was in the README and I am just getting nothing in return. I had some warnings running with my M1 Mac, so I modified the command to see if that would help. I also tried to set environment variables to put Node into a debug mode to see if I can see the request.

### Command I am running
I am running this in the directory of my README file on my Mac.
```bash
docker run --platform linux/amd64 -v $(pwd):/workspace \
  -e DOCKERHUB_USERNAME='mydockerhubusername' \
  -e DOCKERHUB_PASSWORD='mysupersecretpat' \
  -e DOCKERHUB_REPOSITORY='serversideup/php' \
  -e README_FILEPATH='/workspace/README.md' \
  -e NODE_ENV=dev \
  -e DEBUG=* \
  peterevans/dockerhub-description:3
  ```

### My results after running command above
The command runs and returns no output. When I run `echo $?`, it returns a `1` -- hinting there is an error. It's challenging to know what that is since there isn't any output being displayed.

<img width="910" alt="Screenshot 2023-02-14 at 15 15 19" src="https://user-images.githubusercontent.com/3174134/218865104-3eb4fa08-e3fe-401d-af72-d1d71631636b.png">

### Things I have checked
- ✅ Reset PAT
- ✅ GitHub Actions secrets are reset and named correctly

Thanks for your help everyone! 🙌

**Comment by jaydrogers** [comment]
Yup! That's definitely it. Sorry for the confusion! I thought it had something to do with my PAT.

Weird thing is, if I copy and paste it into hub.docker.com, it works no problem.

I even tried pasting it into hub.docker.com then copying that into my README and it still would not take it.

If there are any quick thoughts to trace the down the culprit, I am open ears -- but I am working on a documentation site anyways so it will greatly simplify my README in the future anyways.

I can open another issue on it in the future if I run into problems.

Thanks for your help 🙌

**Comment by simmel** [comment]
Report this to Docker Hub (somehow). It's their API saying "nah, user
input is wrong".

Good job finding the problem!

**Comment by christian-korneck** [comment]
> If there are any quick thoughts to trace the down the culprit, I am open ears

@jaydrogers do you get the same error when you try to upload your repo description using [this](https://github.com/christian-korneck/docker-pushrm) tool?

**Comment by LeviPesin** [comment]
I'm still getting the "Forbidden" error on the PATCH request. I triple-checked all repository names and even tried running exactly the same API requests in Python -- but no matter what I keep getting 403 `{'message': 'insufficient scope', 'errinfo': {}}` errors.

**Comment by peter-evans** [comment]
@LeviPesin As the error message suggests, the problem is likely to be insufficient scopes. It requires `read/write/delete` scopes (Admin level) to update the Docker Hub repo description. This is mentioned in the readme [here](https://github.com/peter-evans/dockerhub-description#action-inputs).

**Comment by LeviPesin** [comment]
Ah, sorry, indeed!! My token only had read/write permissions. Thank you!
